src/viewport/index.tsx

```diff
@@ -547,66 +543,36 @@ class Viewport extends React.Component<
         }
     }
 
-    public changeAgentColor(
-        agents: string | SelectionEntry[],
-        color: string
-    ): void {
-        if (typeof agents === "string") {
-            agents = this.convertAgentNameToSelectionEntry(agents);
-        }
-        const agentIds = this.selectionInterface.getColorChangeAgentIds(agents);
-        const colorId = this.getColorId(
-            color,
-            this.props.agentColors as string[]
-        );
-        this.applyColorToAgents(agentIds, colorId);
+    private getColorId(color: string): number {
+        /**
+         * Check if the new color is in our current array of color options, if not,
+         * add it before returning the index
+         */
+        return this.visGeometry.addNewColor(color);
     }
 
-    public convertAgentNameToSelectionEntry(name: string): SelectionEntry[] {
-        const uiDisplayData = this.selectionInterface.getUIDisplayData();
-        const agentNames = uiDisplayData.map((a) => a.name);
-        if (!agentNames.includes(name)) {
-            throw new Error("Agent name not found");
-        } else {
-            const colorChangeAgents = {
-                name: name,
-                tags: [],
-            };
-            return [colorChangeAgents];
-        }
-    }
+    public changeAgentsColor(colorChanges: ColorChanges): void {
+        const { agents, color } = colorChanges;
+        const agentIds =
+        this.selectionInterface.getAgentIdsByNamesAndTags(agents);
+        this.selectionInterface.updateAgentColors(agentIds, colorChanges);
+        const colorId = this.getColorId(color);
+        this.visGeometry.applyColorToAgents(agentIds, colorId);
 
-    public getColorId(color: string, agentColors: string[] | number[]): number {
-        const hexColorCodeRegex = /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/;
-        if (!hexColorCodeRegex.test(color)) {
-            throw new Error("Invalid color code");
-        }
-        const typeCastAgentColors = agentColors as string[];
-        if (!typeCastAgentColors.includes(color)) {
-            const updatedColors = [...agentColors, color] as string[]
-            this.setColors(updatedColors);
-        }
-        const colorId = typeCastAgentColors.indexOf(color);
-        return colorId;
     }
 
     public setColors(agentColors: string[] | number[]): void {
-         this.visGeometry.clearColorMapping();
-         const uiDisplayData = this.selectionInterface.getUIDisplayData();
-         const updatedColors = this.selectionInterface.setAgentColors(
```

Comment:
this really updates all the colors at once, which we don't need to do when we're just changing one agents color. Plus this is relying on a prop staying constant (the agent colors), which is not something this component has control over, because the parent component could just change that whole list. In which case we would want to trigger re-mapping of all the colors